src/modules/sitemap/services/sitemap.service.ts

```diff
@@ -0,0 +1,207 @@
+
+		// Generate itemDetail xml
+		await Promise.all(
+			itemDetailPagesArray.map(async (pages: SitemapItemInfo[], index) => {
+				const renderedXml = `<?xml version="1.0" encoding="UTF-8"?>
+				<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
+					${pages.map(this.renderPage).join('\n')}
+				</urlset>
+				`;
+				xmlUrls.push(await this.uploadXml(renderedXml, `itemDetail${index}`));
+			})
+		);
+
+		// Generate index xml
+		const indexXml = `<?xml version="1.0" encoding="UTF-8"?>
+		<sitemapindex xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
+			${xmlUrls.map(this.renderIndexPage).join('\n')}
+		</sitemapindex>
+		`;
+		await this.uploadXml(indexXml, `index`);
+	}
+
+	public async getSitemapConfig(): Promise<SitemapConfig> {
+		try {
+			const { app_config: config } = await this.dataService.execute<GetSitemapConfigQuery>(
+				GetSitemapConfigDocument
+			);
+
+			return config[0];
+		} catch (err) {
+			throw new InternalServerErrorException('Failed getting sitemap config', err);
+		}
+	}
+
+	public async getContentPagesPaths(): Promise<string[]> {
+		try {
+			const contentPages = await this.contentPagesService.fetchContentPages(
+				0,
+				50000,
+				'title',
+				'asc',
+				'',
+				{ content_type: { _eq: 'PAGINA' } }
+			);
+			const paths = contentPages[0].map((cp) => cp?.path);
+			return paths;
+		} catch (err) {
+			throw new InternalServerErrorException('Failed getting all the content pages', err);
+		}
+	}
+
+	// Helpers
+	// ------------------------------------------------------------------------
+	private renderPage(pageInfo: SitemapItemInfo): string {
+		return `<url>
+			<loc>${pageInfo.loc}</loc>
+			${pageInfo.lastmod ? `<lastmod>${pageInfo.lastmod}</lastmod>` : ``}
+			<changefreq>${pageInfo.changefreq}</changefreq>
+			${pageInfo.priority ? `<priority>${pageInfo.priority}</priority>` : ``}
+			</url>`;
+	}
+
+	private renderIndexPage(url: string): string {
+		return `<sitemap>
+			<loc>${url}</loc>
+			</sitemap>`;
+	}
+
+	private async uploadXml(xml: string, name: string): Promise<string> {
+		return await this.assetsService.upload(AssetFileType.SITEMAP, {
+			fieldname: name,
+			originalname: name,
+			encoding: '',
+			mimetype: 'text/xml',
+			size: 0,
+			stream: new Readable(),
+			destination: '',
+			filename: name,
+			path: '',
+			buffer: Buffer.from(xml, 'utf-8'),
+		});
+	}
+
+	private blacklistAndPrioritizePages(
+		pages: SitemapItemInfo[],
+		config: SitemapConfig
+	): SitemapItemInfo[] {
+		const configPaths = config.value.map((c) => c.path);
+
+		return pages.reduce((result, page) => {
+			const path = page.loc.substring(process.env.CLIENT_HOST.length);
+			if (configPaths.includes(path)) {
+				const configValue = config.value.find((c) => c.path === path);
+				if (configValue?.blacklisted === true) {
+					return result;
+				} else {
+					return [...result, { ...page, priority: configValue?.priority }];
+				}
+			}
+			return [...result, page];
+		}, []);
```

Comment:
I'm not a big fan of reduce since i find it hard to see what is going on. I prefer to use compact(....map(...))
compact removes all null/undefined values from an array

you don't need to compare with true inside an if statement

We can also easily set the priority on the page even though it wasn't a requirement.

statements like this: `[...result, page];` can get quite expensive when you use them in a for loop and on large arrays (50000). Since every loop you're asking the pc to copy all values into a new array. Which causes a lot of memory copying. After the reduce you'll have copied 2.500.000.000 page items. 50.000 * 50.000

```suggestion
		return compact(pages.map((page) => {
			const path = page.loc.substring(process.env.CLIENT_HOST.length);
			if (configPaths.includes(path)) {
				const configValue = config.value.find((c) => c.path === path);
				if (configValue?.blacklisted) {
					return null;
				}
				if (configValue?.priority) {
					page.priority = configValue?.priority;
				}
			}
			return page;
		}));
```